drivers/pwm/pwm_led_esp32.c

```diff
@@ -0,0 +1,517 @@
+
+	pwm_led_esp32_low_speed_update(speed_mode, channel);
+}
+
+static int pwm_led_esp32_channel_set(int pin, bool speed_mode, int channel,
+				     int duty, int timer)
+{
+	const int pin_mode = GPIO_DIR_OUT;
+
+	const char *device_name;
+	struct device *gpio;
+	int ret;
+	u32_t sig_out_idx;
+
+	/* Set duty cycle */
+	pwm_led_esp32_duty_set(speed_mode, channel, duty);
+
+	pwm_led_esp32_update_duty(speed_mode, channel);
+
+	pwm_led_esp32_bind_channel_timer(speed_mode, channel, timer);
+
+	/* Set pin */
+	device_name = esp32_get_gpio_for_pin(pin);
+	if (!device_name) {
+		return -EINVAL;
+	}
+
+	gpio = device_get_binding(device_name);
+	if (!gpio) {
+		return -EINVAL;
+	}
+
+	ret = gpio_pin_configure(gpio, pin, pin_mode);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (speed_mode == PWM_LED_ESP32_HIGH_SPEED) {
+		sig_out_idx = LEDC_HS_SIG_OUT0_IDX + channel;
+	} else {
+		sig_out_idx = LEDC_LS_SIG_OUT0_IDX + channel;
+	}
+	esp32_rom_gpio_matrix_out(pin, sig_out_idx, 0, 0);
+
+	return 0;
+}
+
+static int pwm_led_esp32_timer_set(int speed_mode, int timer,
+				   int bit_num, int frequency)
+{
+	u32_t timer_addr;
+	u64_t div_num;
+	int tick_sel = PWM_LED_ESP32_APB_CLK_FREQ;
+	u32_t precision = (0x1 << bit_num);
+
+	/* This expression comes from ESP32 Espressif's Technical Reference
+	 * Manual chapter 13.2.2 Timers.
+	 * div_num is a fixed point value (Q10.8).
+	 */
+	div_num = ((u64_t) APB_CLK_FREQ << 8) / frequency / precision;
+
+	if (div_num < 0x100) {
+		/* Since Q10.8 is a fixed point value, then div_num < 0x100
+		 *  means divisor is too low.
+		 */
+		return -EINVAL;
+	}
+
+	if (div_num > 0x3FFFF) {
+		/* Since Q10.8 is a fixed point value, then div_num > 0x3FFFF
+		 * means divisor is too high. We can try to use the REF_TICK.
+		 */
+		div_num = ((u64_t) 1000000 << 8) / frequency / precision;
+		if (div_num < 0x100 || div_num > 0x3FFFF) {
+			return -EINVAL;
+		}
+		tick_sel = PWM_LED_ESP32_REF_TICK_FREQ;
+	} else {
+		if (speed_mode) {
+			sys_set_bit(LEDC_CONF_REG, LEDC_APB_CLK_SEL_S);
+		}
+	}
+
+	if (speed_mode == PWM_LED_ESP32_HIGH_SPEED) {
+		timer_addr = PWM_ESP32_HSTIMER(timer);
+	} else {
+		timer_addr = PWM_ESP32_LSTIMER(timer);
+	}
+
+	set_mask32(div_num << LEDC_DIV_NUM_LSTIMER0_S, timer_addr);
+	set_mask32(tick_sel << LEDC_TICK_SEL_LSTIMER0_S, timer_addr);
+	set_mask32(bit_num & LEDC_LSTIMER0_LIM_M, timer_addr);
+
+	if (speed_mode) {
+		/* update div_num and bit_num */
+		sys_set_bit(timer_addr, LEDC_LSTIMER0_PARA_UP_S);
+	}
+
+	/* reset low speed timer */
+	sys_set_bit(timer_addr, LEDC_LSTIMER0_RST_S);
+	sys_clear_bit(timer_addr, LEDC_LSTIMER0_RST_S);
+
+	return 0;
+}
+
+/* period_cycles is not used, set frequency on menuconfig instead. */
+static int pwm_led_esp32_pin_set_cycles(struct device *dev,
+					u32_t pwm, u32_t period_cycles,
+					u32_t pulse_cycles)
+{
+	ARG_UNUSED(period_cycles);
+
+	int speed_mode;
+	int channel;
+	int timer;
+	int ret;
+
+	const struct pwm_led_esp32_config * const config =
```

Comment:
Style nitpick: please declare all the variables, without vertical space between them, and then use things like ARG_UNUSED().  Avoid declaring variables outside the declaration of a new scope.